common/io.hpp

```diff
@@ -155,11 +155,11 @@ class StringReader : virtual public Reader {
 	unique_ptr<StreamReader> reader_;
 
 public:
-	StringReader(string &str) :
+	StringReader(const string &str) :
 		s_ {str},
 		reader_ {new StreamReader(s_)} {
 	}
-	StringReader(string &&str) :
+	StringReader(const string &&str) :
```

Comment:
I think const rvalue reference is wrong. It is supposed to modify the original, but the const prevents that. The `stringstream` constructor with rvalue reference is anyway C++20, so it wouldn't have been used anyway, but it is indeed non-const, and this needs to be non-const too, if it is ever to be used.

Reply:
`stringstream` with `rvalue` ref is `C++11` according to `cppreference`: 

https://en.cppreference.com/w/cpp/io/basic_stringstream/basic_stringstream

And I'm not following, how is it supposed to modify the original? It's a reader?